engine/crates/engine-v2/src/plan/collected.rs

```diff
@@ -0,0 +1,83 @@
+use schema::{DataType, FieldId, Wrapping};
+
+use crate::{
+    request::{BoundFieldId, FlatTypeCondition, SelectionSetType},
+    response::{ResponseEdge, ResponseKey},
+    utils::IdRange,
+};
+
+use super::{CollectedFieldId, CollectedSelectionSetId, ConditionalFieldId, ConditionalSelectionSetId, PlanBoundaryId};
+
+#[derive(Debug)]
+pub enum AnyCollectedSelectionSet {
+    /// Generated during planning
+    Collected(CollectedSelectionSetId),
+    Conditional(ConditionalSelectionSetId),
+    /// Generated at runtime from conditional selection sets
+    RuntimeMergedConditionals {
+        ty: SelectionSetType,
+        selection_set_ids: Vec<ConditionalSelectionSetId>,
+    },
+    RuntimeCollected(Box<RuntimeCollectedSelectionSet>),
+}
+
+/// Selection set that could not be entirely collected because of type conditions, we need to know
+/// the actual type before collecting the fields.
+#[derive(Debug, Clone)]
+pub struct ConditionalSelectionSet {
+    // needed to know where to look for __typename
+    pub ty: SelectionSetType,
+    pub maybe_boundary_id: Option<PlanBoundaryId>,
+    pub fields: IdRange<ConditionalFieldId>,
+    pub typename_fields: Vec<(Option<FlatTypeCondition>, ResponseEdge)>,
```

Comment:
You can have:
```
{
  ... on Dog {
   __typename
  }
  myalias: __typename
  __typename
}
```

So it's all the fields for which we need to add the __typename

Reply:
Ok, so `ResponseEdge` is actually just a field in the response?  And `FlatTypeCondition` being none means that it's not inside a type condition?

Can we get some of that in a comment (maybe including your example)

Reply:
yeah. `ResponseEdge` is roughly just a field. It's named `edge` because it also includes indices for the error path and we're not using `ResponseKey` directly because we encode the position of the fields within the edge. This ensures the `BTreeMap` containing the fields is already in the right order when serializing the response.

Reply:
The structure of `ResponseEdge` and its relation with `BoundResponseKey`/`ResponseKey` is presented here: https://github.com/grafbase/grafbase/blob/brabier/planning-v2/engine/crates/engine-v2/src/response/path.rs